tests/git/utils.ts

```diff
@@ -1,72 +1,270 @@
+  init?: boolean;
+}) {
+  const gitDirs = {
+    fs,
+    dir,
+    gitdir,
+  };
+  const authorDetails = {
     author: {
-      name: 'TestCommitter',
+      name: author,
+      email: `${author}@test.com`,
     },
-    message: 'Initial Commit',
-  });
-  await efs.promises.writeFile(
-    path.join('.git', 'packed-refs'),
-    '# pack-refs with: peeled fully-peeled sorted',
-  );
-  for (let i = 0; i < 10; i++) {
-    const fp = i.toString();
-    await efs.promises.writeFile(fp, 'secret ' + i.toString());
-    await git.commit({
-      fs: efs,
-      dir: '.',
-      author: {
-        name: 'TestCommitter ' + i.toString(),
-      },
-      message: 'Commit ' + i.toString(),
+    committer: {
+      name: author,
+      email: `${author}@test.com`,
+    },
+  };
+  if (init) {
+    await git.init({
+      ...gitDirs,
     });
   }
-  const log = await git.log({
-    fs: efs,
-    dir: '.',
-  });
-  if (packFile) {
-    const pack = await git.packObjects({
-      fs: efs,
-      dir: '.',
-      oids: [...log.map((item) => item.oid)],
-      write: true,
+  for (const { message, files } of commits) {
+    await Promise.all(
+      files.map(({ name, contents }) =>
+        fs.promises.writeFile(path.join(gitDirs.dir, name), contents),
+      ),
+    );
+    await git.add({
+      ...gitDirs,
+      filepath: files.map(({ name }) => name),
     });
-    if (indexFile) {
-      await git.indexPack({
-        fs: efs,
-        dir: '.',
-        filepath: path.join('.git', 'objects', 'pack', pack.filename),
-      });
+    await git.commit({
+      ...gitDirs,
+      ...authorDetails,
+      message,
+    });
+  }
+}
+
+const objectsDirName = 'objects';
+const excludedDirs = ['pack', 'info'];
+
+/**
+ * Walks the filesystem to list out all git objects in the objects directory
+ * @param fs
+ * @param gitDir
+ */
+async function listGitObjects({
+  fs,
+  gitDir,
+}: {
+  fs: FileSystem;
+  gitDir: string;
+}) {
+  const objectsDirPath = path.join(gitDir, objectsDirName);
+  const objectSet: Set<string> = new Set();
+  const objectDirs = await fs.promises.readdir(objectsDirPath);
+  for (const objectDir of objectDirs) {
+    if (excludedDirs.includes(objectDir)) continue;
+    const objectIds = await fs.promises.readdir(
+      path.join(objectsDirPath, objectDir),
+    );
+    for (const objectId of objectIds) {
+      objectSet.add(objectDir + objectId);
     }
   }
-  return log;
+  return [...objectSet];
 }
 
-async function getPackID(efs: EncryptedFS): Promise<string> {
-  const pack = (
-    await efs.promises.readdir(path.join('.git', 'objects', 'pack'))
-  )[0];
-  return (pack as string).substring(5, 45);
+type NegotiationTestData =
+  | {
+      type: 'want';
+      objectId: string;
+      capabilityList: CapabilityList;
+    }
+  | {
+      type: 'have';
+      objectId: string;
+    }
+  | {
+      type: 'SEPARATOR' | 'done' | 'none';
+    };
+
+function generateTestNegotiationLine(data: NegotiationTestData, rest: Buffer) {
```

Comment:
Can we get some docblocks for this. Also `generateTest...` sounds a bit weird?

Reply:
Added some clarification for it.